pymatgen/entries/compatibility.py

```diff
@@ -407,58 +408,82 @@ def __str__(self):
 
         Returns:
-            An adjusted entry if entry is compatible, otherwise None is
-            returned.
+            ({correction_name: value})
+
+        Raises:
+            CompatibilityError if the entry is not compatible
         """
-        try:
-            corrections = self.get_corrections_dict(entry)
-        except CompatibilityError:
-            return None
-        entry.correction = sum(corrections.values())
-        return entry
+        return
 
-    def get_corrections_dict(self, entry):
+    def process_entry(self, entry):
         """
-        Returns the corrections applied to a particular entry.
+        Apply the energy adjustment to a single entry.
 
-        Args:
-            entry: A ComputedEntry object.
+        If an Entry is not compatible with the selected scheme (e.g., wrong
+        functional or POTCAR) then None should be returned.
 
         Returns:
-            ({correction_name: value})
+            ComputedEntry: An adjusted entry or None
         """
-        corrections = {}
-        for c in self.corrections:
-            val = c.get_correction(entry)
-            if val != 0:
-                corrections[str(c)] = val
-        return corrections
+        if self.clean:
+            # remove any previously documented corrections from the entry
+            entry.energy_adjustments = []
+
+        # apply the corrections
+        try:
+            corrections = self.get_corrections_dict(entry)
+            for k, v in corrections.items():
+                ea = EnergyAdjustment(v, k, self.as_dict())
+                # Has this correction already been applied?
+                if ea.name in [e.name for e in entry.energy_adjustments]:
+                    if ea.value in [e.value for e in entry.energy_adjustments]:
+                        # we already applied this exact correction. Do nothing.
+                        pass
+                    else:
+                        # we already applied a correction with the same name
+                        # but a different value. Something is wrong.
+                        raise CompatibilityError("Entry {} already has an energy "
+                                                 "adjustment called {}, but its "
+                                                 "value differs from the value of"
+                                                 "{:.3f} calculated here. This"
+                                                 "Entry will be discarded."
+                                                 .format(entry.entry_id,
+                                                         ea.name,
+                                                         ea.value
+                                                         )
+                                                 )
+                else:
+                    # Add the correction to the energy_adjustments list
+                    entry.energy_adjustments.append(ea)
+
+            return entry
```

Comment:
@shyamd if we make `process_entry` an abstract method, how can we enforce that corrections aren't applied twice? See the logic here.

In principle this could go into `process_entries`, but then we could have a situation where `process_entry` gives you a different result than `process_entries`.

Reply:
I think you're confusing how `CorrectionList` works with `Compatibility`. Things that are specifically an issue in `CorrectionList` get taken care of there. This also becomes less of a problem if `Compatibility`'s don't become user composable: Apply one then any other and have them add up.

Reply:
@shyamd not sure I follow completely. The issue of corrections being applied twice is not specific to `CorrectionList`. I want to enforce, for any `Compatibility`, that once an entry is processed, it will not be processed again (or it will only be processed again if the previous correction is removed). You are right that the new infrastructure will handle adding corrections from multiple `Compatibility`, but there is nothing in that paradigm to prevent the SAME correction being added multiple times.

For example, (in pseudo code), I want the abstract class to guarantee that this situation won't happen:
```
>>> compat = Compatibility()
>>> entry = ComputedEntry()
>>> entry.correction
0
>>> compat.process_entry(entry)
>>> entry.correction
2
>>> compat.process_entry(entry)
>>> entry.correction
4
>>> compat.process_entry(entry)
>>> entry.correction
6
```

Related to this, I want to enforce that any `Compatibility` can accept a `clean` argument to specify whether or not previously-applied corrections are removed before processing. This would apply to ALL corrections on the entry, not just those added by the `Compatibility` class that is doing the processing.

For example (pseudo code):
```
>>> entry = ComputedEntry(... correction=2)
>>> entry.energy_adjustments
Manual adjustment: 2
>>> compat = Compatibility(clean=False)
>>> compat.process_entry(entry)
>>> entry.energy_adjustments
Manual adjustment: 2
Example energy adjustment: 5
>>> entry.correction
7
```
vs

```
>>> entry = ComputedEntry(... correction=2)
>>> entry.energy_adjustments
Manual adjustment: 2
>>> compat = Compatibility(clean=True)
>>> compat.process_entry(entry)
>>> entry.energy_adjustments
Example energy adjustment: 5
>>> entry.correction
5
```

If there is a way we can enforce these behaviors while eliminating `get_corrections_dict`, I'm all for it. I just haven't been able to figure out how to do that. Open to ideas!

Reply:
I think you're trying too hard to enforce something that is expected by whoever writes `process_entry`. You have to expect that whoever writes that knows what they are doing and in all-reality, might have a good reason for having two of the same correction. 

One thing you can enforce is that all corrections, except `ManualAdjustment` get stripped before going into `process_entry`

Reply:
Fair point, future developers will know what they're doing so to some extent we can trust them to implement these things. But doesn't this compromise the objective of defining a consistent interface for users (who may NOT know what they are doing)? 

If we don't enforce this, then the behavior of `process_entries` will be ambiguous. In some cases, it could be something that happens one time (sort of a boolean, like is this entry processed or not), but in others, it might be something that I can do over and over and keep changing the energy.

I think it's really important that we enforce one behavior or another in order to prevent unexpected behavior / bugs (this has tripped me up multiple times)

Reply:
> One thing you can enforce is that all corrections, except `ManualAdjustment` get stripped before going into `process_entry`
> 
> Edit: This would be in `process_entries`.

True, but if I do this, then `process_entry` could give me a different answer than `process_entries`. That's going to violate users expectations. If I process an entry with `process_entry` or `process_entries`, I need to wind up with the same energy.

Reply:
Would that be the case for a `Scan` compatibility? I think what you're getting at is that `process_entry` is public so a user might use it; and I'm suggesting that maybe that's not the case. `process_entry` is an implementation method that gets used by `process_entries`.  `process_entries` is used in the MPRester and I think most of the example notebooks.

Reply:
You can even rename it to `_process_entry` if you want so that it's naming is that it is for internal purposes. @mkhorton  how do you feel about that?

Reply:
FYI, `process_entry` is used in three places in `emmet-old`: https://github.com/materialsproject/emmet/search?q=process_entry&unscoped_q=process_entry

Reply:
I'm ok with an explicitly private `_process_entry` being the abstract method and `process_entries` being public. I still think there is some risk of confusion if `_process_entry` doesn't do exactly the same thing that `process_entries` does, but we can manage that with docstrings, etc. 

The one caveat is that I want to allow `process_entries` to take in either a list of entries or a single entry, so that users can work on single entries without having to pass `[entry]`.

If this sounds good to you @shyamd and @mkhorton I'll proceed with that implementation.

Reply:
You can also call it something else from `process_entry`, but basically there should be one interface at the `Compatability` level that users use to get entries adjusted and one interface on the implementation side on where people put their code. Since we want to keep this as integrated with existing pymatgen as possible, it's important to keep `process_entries`. 

Having `process_entries` deal with both single and list of entries is perfectly reasonable.

Reply:
It's important that `process_entries` return be consistent though. It should always return a list of adjusted entries.

Reply:
Good morning guys. After sleeping on it I want to make one more counterproposal, because I'm still concerned about inconsistency between `_process_entry` and `process_entries` causing confusion for users and problems for maintenance / debugging.

What about the following, in which `adjust_energy` is the abstract method?
```
@abstractmethod
adjust_energy(Entry) -> [EnergyAdjustment]

_process_entry(Entry) -> Entry
    calls adjust_energy 

process_entries(Entry or [Entry]) -> [Entry]
    calls _process_entry 
```

This way we have only two public interfaces as @shyamd suggests, we enforce de-duplication and the `clean` argument in `_process_entry`, and we maintain complete consistency between `_process_entry` and `process_entries`.

Compared to what's shown now, I'm proposing to replace `get_corrections_dict` with `adjust_energy` that actually returns `EnergyAdjustment` instead of just a dict, and then making `process_entry` private.

Thoughts @shyamd and @mkhorton ?

Reply:
I'm mostly fine with this. There is no point in having `process_entry` be private then. It's a fully functioning public method. The point of making something private is not to hide it, but to ensure it only gets used in specified situations or manners.

Reply:
As I look at it, I suppose we could also eliminate `_process_entries` entirely in this implementation. I think my main concern is avoiding having two methods (private or not) that look like they do the same thing but have subtle differences. `_process_entry` and `process_entries` "look" like they should behave exactly the same way, just operate on different numbers of `Entry`. But we could just have `get_adjustments` and `process_entries`. What about that?

Reply:
That way, in addition to the semantic difference in the method names, they also clearly have different functions. One returns a list of adjustments, the other modifies the entry.

Reply:
That's fine with me.